src/Control/HTTPApplication.php

```diff
@@ -141,4 +152,48 @@ public function execute(HTTPRequest $request, callable $callback, $flush = false
             $this->getKernel()->shutdown();
         }
     }
+
+    /**
+     * Trigger deprecation notices for legacy configuration which is deprecated but
+     * doesn't have deprecation notices directly on the relevant API
+     *
+     * Don't remove this method even if it's just a no-op - we'll reuse this mechanism
+     * in the future as needed.
+     */
+    private function warnAboutDeprecatedSetups()
+    {
+        // TypeCreator is a class unique to GraphQL v3 - we use it in other areas to detect
+        // which version is being used.
+        if (class_exists(TypeCreator::class)) {
+            Deprecation::notice(
+                '4.13.0',
+                'silverstripe/graphql 3.x is deprecated. Upgrade to 4.x instead.'
+                . ' See https://docs.silverstripe.org/en/4/upgrading/upgrading_to_graphql_4/',
+                Deprecation::SCOPE_GLOBAL
+            );
+        }
+
+        // The alternate_public_dir config property is deprecated, but because it's
+        // always fetched it'll throw a deprecation warning whether you've set it or not.
+        // There are also multiple mechanisms which can result in this bad configuration.
+        if (PUBLIC_DIR !== 'public' || Director::publicDir() !== PUBLIC_DIR) {
+            Deprecation::notice(
+                '4.13.0',
+                'Use of a public webroot other than "public" is deprecated.'
+                . ' See https://docs.silverstripe.org/en/4/changelogs/4.1.0#public-folder/',
+                Deprecation::SCOPE_GLOBAL
+            );
+        }
+
+        // This change of defaults has no other deprecation notice being emitted currently.
+        $project = new Module(BASE_PATH, BASE_PATH);
+        if (RESOURCES_DIR !== '_resources' && $project->getResourcesDir() !== RESOURCES_DIR) {
```

Comment:
This is confusing me.  The PUBLIC_DIR if statement above uses `||` which is intuitive.   However the if statement for RESOURCES_DIR uses `&&` which is not intuitive.

Reply:
Also the logic in Module::getResourcesDir() is
```php
        return isset($this->composerData['extra']['resources-dir'])
            ? $this->composerData['extra']['resources-dir']
            : '';
```
So, are we deprecating NOT specifying what the resources dir is in composer.json?  Just having a quick look at a composer.json for a cms 5 install I have locally and I don't have it specified here and everything works normally.

Seems like we should be doing the opposite and simply not care what's in composer.json, cos it'll always be _resources going forward?

Reply:
> The PUBLIC_DIR if statement above uses || which is intuitive. However the if statement for RESOURCES_DIR uses && which is not intuitive.

These conditions are doing two different things.

```php
if (PUBLIC_DIR !== 'public' || Director::publicDir() !== PUBLIC_DIR) {
```
If `PUBLIC_DIR` isn't public, we have a problem. If `Director::publicDir()` is giving a value other than `PUBLIC_DIR`, we have a problem.

```php
if (RESOURCES_DIR !== '_resources' && $project->getResourcesDir() !== RESOURCES_DIR) {
```
`RESOURCES_DIR` is allowed to be "_resources", and `$project->getResourcesDir()` is allowed to not match `RESOURCES_DIR` (e.g. if not set) - but if `RESOURCES_DIR` isn't "_resources" _then_ `$project->getResourcesDir()` must match that value. If it doesn't, it means they're probably using the _old default_ value and haven't explicitly declared they want to keep it.

Reply:
> So, are we deprecating NOT specifying what the resources-dir is in composer.json?
Only if they're using something other than the default, but yes.

Reply:
RESOURCES_DIR gets set in constants.php, where it'll read composer.json resources-dir or default to `resources` (CMS 4) or `_resources` (CMS 5)

Maybe I'm reading this wrong, though maybe all we need here is

`if ($project->getResourcesDir() === '')`

Meaning you haven't specified it in composer.json and are still using the old 'resources' default?  But is that actually a problem? If you don't specify it in CMS 5 and it'll nicely use the _resources default.

So, people will be upgrading from 4.13 to 5.0.0, they turn on deprecation notices, and we then asking people to specifying resources-dir ... only to then not require it in CMS 5?

What happens when you upgrade from CMS 4 -> CMS 5 and you haven't specified resources-dir in composer.json?  Does the old resources dir get nuked and a new _resources dir get populated?

I have some feeling that not specifying _resources in composer.json causes them to break right now anyway

I'm not sure we need this deprecation warning here? Maybe we need to split this off to another card here to explore this further?

Reply:
> Maybe I'm reading this wrong, though maybe all we need here is `if ($project->getResourcesDir() === '')`

While that would technically work in CMS 4 it's not as clear in its intention, since in CMS 5 you _don't_ need to declare it in composer to use the new default.
The way it's written is intendeed to make it clear what the intention of the deprecation warning is - i.e. "You aren't using the new default, AND you haven't explicitly stated you don't want to use it"

I'll change it to that simplified version if you insist since it does the same thing... but _it does the same thing_ so I don't think it's worth holding up the PR for it.

> Meaning you haven't specified it in composer.json and are still using the old 'resources' default? But is that actually a problem? If you don't specify it in CMS 5 and it'll nicely use the _resources default.
> So, people will be upgrading from 4.13 to 5.0.0, they turn on deprecation notices, and we then asking people to specifying resources-dir ... only to then not require it in CMS 5?

Right... but then it's looking in `_resources` for all your resources, but your project still has `public/resources/` dir where all your stuff is, and it's possible you've even got some hardcoded stuff expecting that directory which will break without you having ever been warned (beyond a short line in a changelog that could be easy to miss amongst all the other changes).

The correct course of action in an upgrade for a project still using `resources` is to either explicitly set `extra.resources-dir` to "resources" or to set it to `_resources` while you're in CMS 4 and make sure that all works for you. The deprecation notice makes it clear action is required, so you can do these steps, and when they're done you're either using `_resources` or you've explicitly declared another folder, and the deprecation warning disappears.

> What happens when you upgrade from CMS 4 -> CMS 5 and you haven't specified resources-dir in composer.json? Does the old resources dir get nuked and a new _resources dir get populated?

I expect the old dir stays around and a new one is populated. But I don't think it's worth investigating. Using the old directory without explicitly stating you want to use it is deprecated. That's what this message informs the developer.

Reply:
I think the simplified `if ($project->getResourcesDir() === '')` has much better intent, because if a project turn on deprecation warnings in 4.13 we're basically saying to them "You must set extra.resources-dir in composer.json to something".  If they set it to _resources then then can remove it in CMS 5, if they don't then it needs to stay there in CMS 5.

The message in the Deprecation::notice() needs to be changed tell users that they need to set this value in composer.json, preferably to _resources, but realistically resources to reduce upgrade pain?

Reply:
Just a note on the `_resources` vs `resources`. There's been conflicts with pages named Resources reported by CMS users which is how/why the underscore came to be, so we shouldn't exercise the option of coming back to `resources`.

Reply:
> we shouldn't exercise the option of coming back to resources

Not sure what you mean there... do you mean we shouldn't tell people they can set that value? Or that we should explicitly _not_ allow it in CMS 5? Or.... something else?

Reply:
It was a reaction to Steve's "the message ... needs to tell users they need to set the value..." where it seemed to me that suggesting `resources` could be there to reduce upgrade pain.

Offer its configurability via composer.json, suggest `_resources`, and discourage `resources` (if needs to be mentioned at all).

Reply:
> I think the simplified if ($project->getResourcesDir() === '') has much better intent

I disagree, but it's _really_ not worth holding up the PR for this so I've made the change.

> The message in the Deprecation::notice() needs to be changed tell users that they need to set this value in composer.json, preferably to _resources, but realistically resources to reduce upgrade pain?

I don't think it does. It includes a link to the changelog, which is the source of truth for whatever advice we want to give about how to handle this scenario. If you think the information in the changelog is insufficient please note what specific changes should be made in the changelog (note it in the issue since you've already merged the docs PR) and I'll update the changelog.

Reply:
Oh yeah I remember the "pages named resources" thing from years ago.  Yeah, really bad to have it called 'resources' because it's a common name for a page, would be ideal if anyone still on that migrated off that.

> we shouldn't exercise the option of coming back to resources

Yeah that would be great, should just force it to be _resources and that's it, but probably too late now that we're in the beta.  

The changelog is a bit confusing in this context because they're on a 4.13 project, but the changelog is for 5.0.0.  In 4.13 we ideally want them to add "_resources" to their composer.json, though in the 5.0.0 changelog we say "If your composer.json file has its extra.resources-dir key set to _resources, you can remove that now." ... so that's confusing

I'd like to just make the message be "add _resources to your composer.json now" - but I've got no idea what happens when you do that - we probably should split this off as a new card to investigate the amount of friction there is around the resources folder

Reply:
> we probably should split this off as a new card to investigate the amount of friction there is around the resources folder

Whether the readme needs to be updated or further investigation is required, that should not hold up this PR.

Reply:
Yeah I'll raise a card to investigate further, keep as is for now